File: VisualStudioFiles/FEBIO_PLUGIN/fiberConstEqn.cpp
/* fiberConstEqn.cpp ------------------------------------------------------------------------------

SUMMARY: fiberConstEqn.cpp is a function with several fiber constitutive equations implemented.
To change the fiber constitutive equation simply change the fibertype of fibers in the network or
change the corresponging fiber const eqn in this function.
This function also calculates dF/dlambda for the constitutive equations for use in the analytical 
Jacobian calculation. 

CALLED FROM: calcForcesPeriodic.cpp, calcJacobian2.cpp
CALLS ON: None

INPUTS: vect1 - 3x1 vector along fibers
		node_force1 - 3x1 vector of forces acting on node1 of fiber
		node_force2 - 3x1 vector of forces acting on node2 of fiber
		init_len - double value of fiber initial length
		fibtype - int value of fiber type
		fib_area - double value of fiber cross-sectional area
		fib_force - double value of fiber force
		dFdlam - double value of dF/dlambda for the constitutive model

OUTPUTS: None

NOTE: node_force1, node_force2, fib_force, and dFdlam are modified by this code.

Author: Ryan Mahutga - Alford Lab - University of Minnesota

Last Update: 04-10-2023

*/

#include <iostream>
#include "Eigen/Eigen"
#include "calcForcesPeriodic.h"

using namespace Eigen;
using namespace std;

const double PI = 3.141592653589793238462643383279502884;

void fiberConstEqn(Vector3d vect1, Vector3d& node_force1, Vector3d& node_force2, double& lambda, double init_len, 
	int fibtype,double fib_area, double& fib_force, double& dFdlam)
{
	double GS = 0, // Green Strain
		fib_mod = 0, // fiber modulus
		fib_length = 0; // fiber length

	fib_length = vect1.norm(); // fiber length

	lambda = fib_length / init_len; // fiber stretch

	if (fibtype == 1 || fibtype == 12) // fiber force from helical model in Freed, A.D. & Doehring, T.C. (2005) J. Biomech. Eng. 127  
	{ // 3 is for collagen

			fib_mod = 120e6;

			double R0 = 9.8; // [nm] roughly radius of collagen microfibril
			double r0 = 1.6; // [nm] roughly a collagen triple helix
			double H0 = 67.4; //[nm] d-pattern banding in collagen molecules
		if (fibtype==12)
		{
			R0 = 8.8; // [nm] roughly radius of collagen microfibril
		}
		double L0 = sqrt((2 * PI*R0)*(2 * PI*R0) + H0 * H0);

		double lambda_bar = L0 / H0;

		double E_bar = fib_mod * H0*H0 / (L0*(H0 + (1 + 37 / (6 * PI*PI) + 2 * (L0*L0) / ((PI*r0)*(PI*r0)))*(L0 - H0)));

		double sigma = 0;

		if (lambda <= lambda_bar) // see above paper for more information
		{
			double H = lambda * H0;
			double R = sqrt(L0*L0 - H * H) / (2 * PI);
			double eta = (R*R + H * H) / (L0*H*(1 + 4 * R*R / (r0*r0) + 6 * (20 / 9 + R * R / (r0*r0))*R*R / (H*H)));

			double dHdlam = H0;
			double dRdH = -H / (2 * PI*sqrt((L0*L0 - H * H)));
			double dEtadH = -(H*H + R * R) / ((H*H) * L0*(6 * (R*R) * ((R*R) / (r0*r0) + 20 / 9) / (H*H) + 4 * (R*R) / (r0*r0) + 1)) +
				2 / (L0*(6 * (R*R) * ((R*R) / (r0*r0) + 20 / 9) / (H*H) + 4 * (R*R) / (r0*r0) + 1)) +
				12 * (R*R) * ((H*H) + (R*R))*((R*R) / (r0*r0) + 20 / 9) / ((H*H*H*H) * L0*((6 * (R*R) * ((R*R) / (r0*r0) + 20 / 9) / (H*H) + 4 * (R*R) / (r0*r0) + 1)*(6 * (R*R) * ((R*R) / (r0*r0) + 20 / 9) / (H*H) + 4 * (R*R) / (r0*r0) + 1)));
			double dEtadR = 2 * R / (H*L0*(6 * (R*R) * ((R*R) / (r0*r0) + 20 / 9) / (H*H) + 4 * (R*R) / (r0*r0) + 1)) -
				((H*H) + (R*R))*(12 * (R*R) / ((r0*r0) * (H*H)) + 12 * R*((R*R) / (r0*r0) + 20 / 9) / (H*H) + 8 * R / (r0*r0)) / (H*L0*((6 * (R*R) * ((R*R) / (r0*r0) + 20 / 9) / (H*H) + 4 * (R*R) / (r0*r0) + 1)*(6 * (R*R) * ((R*R) / (r0*r0) + 20 / 9) / (H*H) + 4 * (R*R) / (r0*r0) + 1)));
			double dEtadlam = dEtadH * dHdlam + dEtadR * dRdH*dHdlam;

			if (lambda >= 1)
			{
				sigma = eta * E_bar*(lambda - 1);
				dFdlam = fib_area * (eta*E_bar + E_bar * (lambda - 1)*dEtadlam);
			}
			else 
			{
				sigma = eta * E_bar*(lambda - 1);
				dFdlam = fib_area * (eta*E_bar + E_bar * (lambda - 1)*dEtadlam);
			}
		}
		else
		{
			sigma = E_bar * (lambda_bar - 1) + fib_mod * (lambda / lambda_bar - 1);
			dFdlam = fib_mod * fib_area / lambda_bar;
		}

			fib_force = fib_area * sigma;
	}
	else if (fibtype == 2 || fibtype==0 || fibtype==5) // fiber force from F = k(GS) where GS = Green strain = 0.5(lambda^2-1) 
	{//(fibtype 2 for elastin, 0 for failed fibers)
		GS = lambda - 1.0;

		if (fibtype == 2)
		{
			if (lambda >= 1.0)
			{
				fib_mod = 1e6; //see Elastic Proteins: biological roles and mechanical properties Gosline, J. et al. (2002) Phil. Trans. R. Soc. Lond.
				dFdlam = fib_mod * fib_area;
			}
			else
			{
				fib_mod = 0.1e6; // 1% of elastin stiffness for failed fibers
				dFdlam = fib_mod * fib_area;
			}
		}
		else if (fibtype == 5) // no tension/compression asymmetry
		{
			fib_mod = 1e6; //see Elastic Proteins: biological roles and mechanical properties Gosline, J. et al. (2002) Phil. Trans. R. Soc. Lond.
			dFdlam = fib_mod * fib_area;
		}
		else if (fibtype == 0) // failed/removed
		{
			if (lambda >= 1.0) 
			{
				fib_mod = 1; 
				dFdlam = fib_mod * fib_area;
			}
			else
			{
				fib_mod = 1; 
				dFdlam = fib_mod * fib_area;
			}
		}
	
		fib_force = fib_mod * fib_area*GS;
	}
	else if (fibtype == 3) // actin
	{
		GS = lambda - 1.0;

		// passive contribution
		if (lambda >= 1.0) //  give 12pN force at lam = 1.05
		{
			fib_mod = 0.5e6; //0.036e9; // 0.0011e9 see Elastic Proteins: biological roles and mechanical properties Gosline, J. et al. (2002) Phil. Trans. R. Soc. Lond.
			dFdlam = fib_mod * fib_area;
		}
		else
		{
			fib_mod = 0.05e6; // 1% of elastin stiffness for compressed
			dFdlam = fib_mod * fib_area;
		}

		fib_force = fib_mod * fib_area * GS;

		//active contribution
		if (lambda > 0.85 && lambda < 1.85)
		{
			double S0 = 0.2e6, l_min = 0.85, l_max = 1.35, WSS_inf = 5.0; // estimate from Alford integrative biology //0.036e9; // 0.0011e9 see Elastic Proteins: biological roles and mechanical properties Gosline, J. et al. (2002) Phil. Trans. R. Soc. Lond.

			double fib_S = S0;

			double f_lam = 1.0 - (l_max - lambda)*(l_max - lambda) / ((l_max - l_min)*(l_max - l_min));

			double fib_force_active = fib_S * f_lam*fib_area;

			double dFdlamS = fib_area * 2.0 * fib_S * (l_max - lambda) / ((l_max - l_min)*(l_max - l_min));

			fib_force = fib_mod * fib_area * GS + fib_force_active;

			dFdlam = dFdlam + dFdlamS;
		}
		else 
		{
			fib_force = fib_mod * fib_area*GS;
		}
	}
	else if (fibtype == 5) // fiber force from F = EA*(exp(GS)-1) where GS = Green strain = 0.5(lambda^2-1)
	{

		double lambda_limit = 3.15;
		double B;

		fib_mod = 1e5;//2569; //  fit to Ruberti single fiber experiment
		B = 2;//77.2;

		GS = 0.5*(lambda*lambda - 1);  // Green Strain

		if (lambda >= lambda_limit)
		{
			GS = 0.5*(lambda_limit*lambda_limit - 1);  // Green Strain at limit
			double force_exp = fib_mod * fib_area*(exp(GS*B) - 1) / B;
			double slope_at_limit = fib_mod * fib_area*lambda_limit*exp(B*GS);
			fib_force = force_exp + slope_at_limit * (lambda - lambda_limit);
			dFdlam = slope_at_limit;
		}
		else if (lambda >= 1)
		{
			fib_force = fib_mod * fib_area*(exp(B*GS) - 1) / B;
			dFdlam = fib_mod * fib_area*lambda*exp(B*GS);
		}
		else if (lambda < 1)
		{
			fib_force = fib_mod * fib_area*(exp(B*GS) - 1) / B;
			dFdlam = fib_mod * fib_area*lambda*exp(B*GS);
		}
	}

	// angles for forces
	Vector3d cosine = vect1 / fib_length;

	// nodal forces due to stretched fibers
	node_force1 = fib_force*cosine;
	node_force2 = -1.0 * fib_force*cosine;

}

File: Matlab Files/Network Generation/FiberConstEqn.m
function [ fib_force, node_1_force, node_2_force, dFdlam ] = FiberConstEqn( vect, init_len, fibtype, fiber_area )
%FiberConstEqn contains the fiber constitutive equation for the periodic
%       RVE solution
%   INPUTS: vect - 3x1 vector from node1 to node2
%           init_len - fiber initial length
%           fibtype - fiber type

%   OUTPUTS: fib_force - force of fiber between nodes
%            node_1_force - 3x1 x,y,z force acting on node1
%            node_2_force - 3x1 x,y,z force acting on node2
%            dFdlam - 1x1 value of the derivative of fiber force with
%            respect to lambda at current lambda

%  Created By: Ryan Mahutga - Barocas Research Group - University of Minnesota
%  Date Created: 02/08/18

a = 2; % 1 is for exponential fibers, 2 is for linear fibers, 3 for helical spring model, 4 for incompresible neo-Hookean fibers

% dealing with input values
x_span = vect(1) ;
y_span = vect(2) ;
z_span = vect(3) ;

fiber_current_length = sqrt( x_span^2 + y_span^2 + z_span^2 ); % calculation of fiber length

lambda = fiber_current_length / init_len; % calculation of fiber stretch

% This is an option to use either exponential or linear fibers -- One can
% adjust the fiber parameters below if desired.

%% Exponential Fiber
if a==1  % EXP FIBER FORCE RELATION F = E A ( exp(B*Green Strain) - 1 ) / B
    
    lambda_limit = 7.5; % fib force linear above fiber green strain of 7.5
    
    % generate fiber modulus and beta value
    if fibtype == 1
        fib_mod = 0.7e9; % gives E = 0.5GPa ( %10.8e6; % gives EA=340nN
        fib_B = 2.5;
    elseif fibtype == 2 % elastin fibers
        fib_mod = 2.5e6;
        fib_B = 2.17;
    elseif fibtype == 3 % radial elastin struts (2x modulus of elastin)
        fib_mod = 2.5e6;
        fib_B = 2.17;
    end
    
    if lambda > lambda_limit
        % LINEAR FORCE ABOVE OUR THRESHOLD STRETCH
        
        gs = 0.5 * ( lambda_limit^2 - 1 );
        
        force_exp = fib_mod * fiber_area * (exp(fib_B * gs)-1) / fib_B;
        
        slope_at_lambda_limit = fib_mod * fiber_area * lambda_limit * exp(fib_B * gs);
        
        force = force_exp + slope_at_lambda_limit * (lambda - lambda_limit);
        
        dFdlam = slope_at_lambda_limit ;
        
    elseif lambda < lambda_limit && lambda >= 1
        
        % REGULAR CONST EXN BELOW THRESHOLD
        
        gs = 0.5 * ( lambda^2 - 1 );
        
        force = fib_mod * fiber_area * (exp( fib_B * gs) - 1) / fib_B;
        
        dFdlam = fib_mod * fiber_area * lambda * exp( fib_B * gs) ;
        
    else
        
        %BELOW LAMBDA = 1 -> use 1e-12 * force
        
        gs = 0.5 * ( lambda^2 - 1 );
        
        force = 1e-12.* fib_mod * fiber_area * (exp( fib_B * gs) - 1) / fib_B;
        
        dFdlam = 1e-12* fib_mod * fiber_area * lambda * exp( fib_B * gs) ;
        
%                 force = 0 ;
        
    end
    
    %% Linear Fiber
elseif a==2  % Linear fiber consitutive equation (F=EA*(Green Strain))
    
    % generate fiber modulus
    if fibtype == 1
        fib_mod = 13e6 ; %10.8e6 ; % gives E = 0.5GPa ( %10.8e6; % gives EA=340nN
    elseif fibtype == 2 % elastin fibers
        fib_mod = 1e6;
    elseif fibtype == 3 % radial elastin struts (2x modulus of elastin)
        fib_mod = 170e6;
    end
    
    if lambda >= 1
        
        gs = 0.5 * ( lambda^2 - 1 );
        
        force = fib_mod * fiber_area * gs ;
        
        dFdlam = fib_mod * fiber_area * lambda ;
        
    else % if lambda < 1 (3 orders of magnitude lower
        
        gs = 0.5*(lambda^2-1); % Green Strain
        
        force = 1e-12.*fiber_area*fib_mod*gs ;
        
        dFdlam = 1e-12 * fib_mod * fiber_area * lambda ;
        
    end
    
%% Helical collagen fibers (Freed, A.D. & Doehring, T.C. (2005) J. Biomech. Eng. 127    
elseif a ==3
    
    % generate fiber modulus
    if fibtype == 1
        E = 0.7e9 ; %10.8e6 ; % gives E = 0.5GPa ( %10.8e6; % gives EA=340nN
    elseif fibtype == 2 % elastin fibers
        E = 2.5e6;
    elseif fibtype == 3 % radial elastin struts (2x modulus of elastin)
        E = 2.5e6;
    end
    
    R0 = 2.9 ;% Diameter of 5.8 nm (roughly a collagen microfibril size)
    r0 = 1.2 ; % Diameter of 2.4 nm (roughly a tropocollagen triple helix)
    H0 = 67 ; % [nm] D band pattern in collagen microfibril
    
    L0 = ((2*pi()*R0)^2 + H0^2)^0.5;
    
    lambda_bar = L0/H0 ;
    E_bar = E*H0^2/(L0*(H0 + (1 + 37/(6*pi()^2) + 2*(L0/(pi()*r0))^2)*(L0-H0))) ;
    
    if lambda <= lambda_bar
        H = lambda * H0 ;
        R = (L0^2-H^2)^0.5/(2*pi()) ;
        eta = (R^2+H^2)/(L0*H*(1 + 4*R^2/r0^2 + 6*(20/9 + R^2/r0^2)*R^2/H^2)) ;
        sigma = eta*E_bar*(lambda-1) ;
        dHdlam = H0 ;
        dRdH = -H/(2*pi())*(L0^2-H^2)^(-0.5) ;
        dEtadH = -(H^2+R^2)/(H^2*L0*(6*R^2*(R^2/r0^2+20/9)/H^2 + 4*R^2/r0^2 + 1)) + ...
            2/(L0*(6*R^2*(R^2/r0^2 + 20/9)/H^2 + 4*R^2/r0^2 + 1)) + ...
            12*R^2*(H^2 + R^2)*(R^2/r0^2 + 20/9)/(H^4*L0*(6*R^2*(R^2/r0^2 + 20/9)/H^2 + 4*R^2/r0^2 + 1)^2) ;
        dEtadR = 2*R/(H*L0*(6*R^2*(R^2/r0^2+20/9)/H^2 + 4*R^2/r0^2 + 1)) -...
            (H^2+R^2)*(12*R^2/(r0^2*H^2) + 12*R*(R^2/r0^2 + 20/9)/H^2 + 8*R/r0^2)/(H*L0*(6*R^2*(R^2/r0^2 + 20/9)/H^2 + 4*R^2/r0^2 + 1)^2) ;
        dEtadlam = dEtadH*dHdlam + dEtadR*dRdH*dHdlam ;
        dFdlam = fiber_area*(eta*E_bar + E_bar*(lambda-1)*dEtadlam) ;
    else
        sigma = E_bar*(lambda_bar - 1) + E*(lambda/lambda_bar - 1) ;
        dFdlam = E*fiber_area/lambda_bar ;
    end
    
    force = fiber_area*sigma ; % estimation since sigma is Cauchy Stress
    
%% Neo-Hookean Incompressible Fiber    
elseif a==4 % this is not working currently 
    % Relation obtained assuming 2 and 3 direction are the same, and that 
    % the strain energy density is given by: W = C1/2*(I-3) where I =
    % lam1^2+lam2^2+lam3^2; (Derived in terms of underformed fiber 
    % x-sectional area.)
    
    % generate fiber modulus
    if fibtype == 1
        fib_mod = 30e6 ; %10.8e6 ; % gives E = 0.5GPa ( %10.8e6; % gives EA=340nN
    elseif fibtype == 2 % elastin fibers
        fib_mod = 2.5e6;
    elseif fibtype == 3 % radial elastin struts (2x modulus of elastin)
        fib_mod = 2.5e6;
    end
    
    if lambda >= 1
        
        force = fib_mod * fiber_area * (lambda - 1/(lambda^2)) ;
        
        dFdlam = fib_mod * fiber_area * (1 + 2*(lambda^-3)) ;
        
    else % if lambda < 1 (3 orders of magnitude lower
        
        force = 1e-12 * fib_mod * fiber_area * (lambda - 1/(lambda^2)) ;
        
        dFdlam = 1e-12 * fib_mod * fiber_area * (1 + 2*(lambda^-3)) ;
        
    end
    
end

%% Preparing Outputs
fib_force = force; % fiber forces

cosine_A = x_span / fiber_current_length;
cosine_B = y_span / fiber_current_length;
cosine_C = z_span / fiber_current_length;


% calculating nodal forces

node_1_force(1) = force * cosine_A * (1);
node_1_force(2) = force * cosine_B * (1);
node_1_force(3) = force * cosine_C * (1);

node_2_force(1) = force * cosine_A * (-1);
node_2_force(2) = force * cosine_B * (-1);
node_2_force(3) = force * cosine_C * (-1);

end

